File: prolog/app/Modules/ValidInput/Validation.hs
module Modules.ValidInput.Validation where

import Data.Char (isDigit)
import Data.List (isInfixOf)
import Model.User

isValidEmail :: String -> Bool
isValidEmail email
  | "@gmail.com" `isInfixOf` email || "@hotmail.com" `isInfixOf` email || "@ccc.ufcg.edu.br" `isInfixOf` email || "@estudante.ufcg.edu.br" `isInfixOf` email || "@outlook.com" `isInfixOf` email || "@yahoo.com" `isInfixOf` email = True
  | otherwise = False

isValidPassword :: String -> Bool
isValidPassword password = length password >= 6

isValidName :: String -> Bool
isValidName name = not (any isDigit name)

isValidIndex :: Int -> [Int] -> Bool
isValidIndex n [] = True
isValidIndex n (x : xs) = (n /= x) && isValidIndex n xs

isValidSize :: [Int] -> Bool
isValidSize [] = True
isValidSize l1 = length l1 < 10

isValidGenre :: String -> Bool
isValidGenre genre = do
  let genreList = words genre
  let validSize = length genreList <= 5
  let validChoices = all (\num -> num >= 1 && num <= 7) (map read genreList :: [Int])
  validSize && validChoices

filterUserList :: String -> [User] -> [User]
filterUserList _ [] = []
filterUserList em (x : xs) =
  if email x == em
    then filterUserList em xs
    else x : filterUserList em xs

File: prolog/app/Modules/ValidInput/Validation.pl
:- module(_,[isValidEmail/2, isValidPassword/2, isValidSize/2, isValidName/1]).
:- use_module("../UtilModule.pl").



isValidName(Name) :-
    string_codes(Name, Codes),
    all_spaces(Codes);
    containsNumber(Name);
    containsSpecialChars(Name).

containsSpecialChars(Name) :-
    string_codes(Name, Codes),
    member(Code, Codes),
    \+ char_type(Code, alpha),
    \+ Code = 32.

containsNumber(Name) :-
    atom_chars(Name, Chars),
    member(Char, Chars),
    char_type(Char, digit).

all_spaces([]).
all_spaces([32|T]) :-  % código ASCII para espaço é 32
    all_spaces(T).

validDomain("gmail.com"):-!.
validDomain("hotmail.com"):- !.
validDomain("ccc.ufcg.edu.br"):- !.
validDomain("estudante.ufcg.edu.br"):- !.
validDomain("outlook.com"):- !.
validDomain("yahoo.com"):- !.


isValidEmail(Email, Result) :-
    split_string(Email, "@", "", [_, Domain]),
    validDomain(Domain),
    Result = 'valido',!.

isValidEmail(_, Result) :-
    Result = 'invalido',!.


isValidPassword(Password, Result):- string_length(Password, Size), Size < 6, Result = 'invalido',!.
isValidPassword(Password,Result):- string_length(Password, Size), Size >= 6, Result = 'valido',!.



isValidIndex(_, []).
isValidIndex(N, [X|Xs]) :-
    N \= X,
    isValidIndex(N, Xs),!.


isValidSize(L, R) :-
length(L, Size),
Size =< 9,  R = 'valido',!.

isValidSize(_, R):- R = 'invalido',!.